### PR TITLE
Matrix4 validation

### PR DESCRIPTION
**Pull request by dnfield**
Helps two issues:

- `SceneBuilder.pushTransform` was using different matrix4 validation logic than painting.dart
- Makes sure that matrices passed in for transformations are finite before getting to the native side.

This would have helped catch https://github.com/flutter/flutter/issues/31650

### CODE REVIEW

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -42,6 +42,9 @@ bool _offsetIsValid(Offset offset) {
 bool _matrix4IsValid(Float64List matrix4) {
   assert(matrix4 != null, 'Matrix4 argument was null.');
   assert(matrix4.length == 16, 'Matrix4 must have 16 entries.');
+  assert(() {
+    return matrix4.every((double value) => value.isFinite);
```
Comment (tvolkert) [inline lib/ui/painting.dart]: No need to wrap this in a closure - it'll get compiled out in AOT mode regardless.

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -42,6 +42,9 @@ bool _offsetIsValid(Offset offset) {
...
+  }(), 'Matrix4 entries must be finite.');
   return true;
```
Comment (chinmaygarde) [inline lib/ui/painting.dart]: You want to perform validation in profile and release modes too. How is passing non-finite matrices to the engine in release mode OK?

File: testing/dart/compositing_test.dart [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      0, 1, 0, 0,
+      0, 0, 1, 0,
+      0, 0, 0, double.nan,
+    ]);
+    expect(
+      () => builder.pushTransform(matrix4NaN),
+      throwsA(const TypeMatcher<AssertionError>()),
+    );
+  });
+}
```
Comment (chinmaygarde) [inline testing/dart/compositing_test.dart]: Nit: Newline at EOF

File: lib/ui/painting.dart (hunk 2 above)
Comment (dnfield) [reply lib/ui/painting.dart]: I'm closing this in favor of @liyuqian's, but we're already doing this all over the place.  Everywhere we assert that a offset, rect, or matrix is valid it's only in debug mode.

**Pull request closed without merging**